### ISSUE TITLE
OpenNLPTokenizer + CharFilter occasionally generates invalid long tokens

### ISSUE THREAD

**Issue by ziqizhang** [ziqizhang/jate#26]
Tested version: JATE 2.0 Alpha
Tested schema:

```
<analyzer type="index">
          <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u0027" replacement=" ' " />
          <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u000a" replacement=" \\n " />
                <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u0009" replacement=" \\t " />
                <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u0008" replacement=" \\b " />
                <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u000d" replacement=" \\r " />
                <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u000c" replacement=" \\f " />
                <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u0022" replacement=" &quot; " />
                <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u005c" replacement=" \\ " />
                <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u003c" replacement=" &lt; " />
                <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u003e" replacement=" &gt; " />
                <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u003d" replacement=" = " />
                <charFilter class="solr.PatternReplaceCharFilterFactory" pattern="\\u0026" replacement=" &amp; " />
                <charFilter class="solr.HTMLStripCharFilterFactory"/>
                <!--tokenizer class="solr.StandardTokenizerFactory" /-->
                <tokenizer class="org.apache.lucene.analysis.jate.OpenNLPTokenizerFactory"
                           sentenceModel="en-sent.bin"
                           tokenizerModel="en-token.bin"/>
                <filter class="solr.ASCIIFoldingFilterFactory"/>
                <filter class="org.apache.lucene.analysis.jate.OpenNLPPOSTaggerFactory"
                        posTaggerClass="uk.ac.shef.dcs.jate.nlp.opennlp.POSTaggerOpenNLP"
                        posTaggerModel="en-pos-maxent.bin"/>
.... <!-- the remaining lines are not important -->
</analyzer>
```

Tested file:
C00-2128, see attachment. Download and change extension to .xml

Description:
When OpenNLPTokenizer is preceded by CharFilters such as that shown above in the schema, it seems that **only on specific files**, OpenNLPTokenizer creates invalid, extremely long token. For example, with the above schema, using the attached file, the following code (beginning from line 114 of OpenNLPTokenizer) will generate a token with 1 character "&", start offset = 7248, end offset=8298.

```
termAtt.setLength(termLength);
char[] buffer = termAtt.buffer();
if(word.getStart()==109 && sentenceOffset==7139)
                System.out.println("wrong");
finalOffset = correctOffset(sentenceOffset + word.getEnd());
int start = correctOffset(word.getStart() + sentenceOffset);
offsetAtt.setOffset(start, finalOffset);
```

But in fact, the end offset should be 7249.

The error seems to be caused by line 4 above, which calls 'correctOffset' of the previous filter in the chain, i.e., HTMLStripCharFilterFactory in the example schema above. Without calling that method, the value "sentenceOffset + word.getEnd()" would have been the correct value.

The invalid offsets of the token can cause issues to subsequent processors in the chain. For example, in OpenNLPPOSTaggerFilter, line 84:

```
word = new String(buffer, 0, offsetAtt.endOffset() - offsetAtt.startOffset());
```

would get StringIndexOutOfBound exception, when it is trying to reconstruct the string using start, end offset on the char array object (buffer) that holds the actual string (1 char length '&'). 

COMMENTS
It has been difficult to debug this issue, because the bug only happens with OpenNLPTokenizer + CharFilter combination, and only on certain files. It does not happen when solr's StandardTokenizer is used, or when tested on a separate input file that also contains HTML entities which will be stripped by HTMLStripCharFilterFactory. For example, the second attached file does not cause this problem.

[C00-2128_cln.change.to.xml.pdf](https://github.com/ziqizhang/jate/files/167444/C00-2128_cln.change.to.xml.pdf)
[C00-2128_cln_small.change.to.xml.pdf](https://github.com/ziqizhang/jate/files/167474/C00-2128_cln_small.change.to.xml.pdf)

**Comment by jerrygaoLondon** [comment]
The problem is still unclear. It seems happens for short-text document that ngram is not properly indexed. charFilter is not the root cause.  A list of documents from ACL RD-TEC 1.0 only containing one short sentence:

P06-1143

C02-1110

C02-1154

C82-1053

W03-0800

P00-1040

C67-1007

C02-1067

C82-2028

C86-1136

...

**Comment by jerrygaoLondon** [comment]
@ziqizhang potential bug in ComplexShingleFilter (/org.apache.lucene.analysis.jate.ComplexShingleFilterFactory) if 'maxTokens' is larger than actual available token size (for the case of document with very short-text)

**Comment by jerrygaoLondon** [comment]
Some of ngrams/candidate terms are also mister not indexed, For example, "acoustic adaptation training" for doc "N01-1017" in ACL RD-TEC 1.0

**Comment by ziqizhang** [comment]
I am closing this issue after some investigation and making some corrections.

1. So far the offset issue only happens when the HTMLStripCharFilterFactory is used, and only when input text contains html entities, such as &quot; This token occupies a 5 character length. But HTMLStripCharFilterFactory will convert this to ". The downstream OpenNLPTokenizer will see the converted character ", which is only a 1-character length, but its recorded offsets will indicate it as a 5-character long token. As a result, it complains that the start, end offsets and the length of the token are not consistent.

This is only a minor issue and should not affect other processors in the pipeline. I have changed the warning message to reflect this. However for now I am keeping that warning message just to remind users to check if this is the case - after all we cannot be assured that JATE is fully tested and that this issue does not occur in other cases.

2. the ComplexShingleFilter had a bug which stops it from creating the right shingles when the input text is very short such that the 'maxTokens' is larger than the actual available token size in the text. This has now been fixed.

Both fix will be included in the next release, which will be available soon and based on Solr 7.2.1.
